### ISSUE TITLE
DeflaterOutputStream won't close its underlying deflater under certain circumstance

### ISSUE THREAD

**Issue by andylizi** [GlowstoneMC/Glowstone#592]
Please excuse any grammar errors I may make, I'm not a native English speaker :)

From [DeflaterOutputStream.java](https://github.com/dmlloyd/openjdk/blob/jdk8u/jdk8u/jdk/src/share/classes/java/util/zip/DeflaterOutputStream.java#L231-L244):
```java
/**
 * Writes remaining compressed data to the output stream and closes the
 * underlying stream.
 * @exception IOException if an I/O error has occurred
 */
public void close() throws IOException {
    if (!closed) {
        finish();
        if (usesDefaultDeflater)
            def.end();
        out.close();
        closed = true;
    }
}
```
The deflater will close only if `usesDefaultDeflater` is set to true, so when a custom deflater was passed to DeflaterOutputStream's constructor, **it'll never be closed** until GC kicks in:
```java
/**
 * Closes the compressor when garbage is collected.
 */
protected void finalize() {
    end();
}
```
Same with [DeflaterInputStream](https://github.com/dmlloyd/openjdk/blob/jdk8u/jdk8u/jdk/src/share/classes/java/util/zip/DeflaterInputStream.java#L119-L138), [InflaterInputStream](https://github.com/dmlloyd/openjdk/blob/jdk8u/jdk8u/jdk/src/share/classes/java/util/zip/InflaterInputStream.java#L218-L230) and [InflaterOutputStream](https://github.com/dmlloyd/openjdk/blob/jdk8u/jdk8u/jdk/src/share/classes/java/util/zip/InflaterOutputStream.java#L174-L190).

This mechanism affects any usage of non-default deflater/inflater, for example [ReginFile#getChunkDataOutputStream(int, int)](https://github.com/GlowstoneMC/Glowstone/blob/a1d7d11948d512ba9d443fee36a67da4d8c41aa3/src/main/java/net/glowstone/io/anvil/RegionFile.java#L267).

**Comment by mastercoms** [comment]
The only solutions I've thought of is making the way to get chunk data streams much more complicated and less standardized or doing a costly reflection operation to set `usesDefaultDeflater` to true, so therefore, it's been like this for now.

The benefits of being able to set a larger buffer size and more optimal deflater outweigh the cost of that extra memory usage from the leak, in my opinion.

**Comment by andylizi** [comment]
thanks for the explanation :)

**Comment by luis-guideti** [comment]
Simple solution:

`        
        DeflaterOutputStream dos = new DeflaterOutputStream(new ChunkBuffer(x, z), deflater, 2048) {
            
            @Override
            public void close() throws IOException {
                super.close();
                def.end(); // supress exceptions here maybe?
            }
        };`

**Comment by mastercoms** [comment]
@FurmigaHumana Hi, would you mind opening a PR with this fix? :)

**Comment by kamcio96** [comment]
Did you think about using Deflater with ThreadLocal? I didn't have time to review code. Will do soon, may I come up with something.

**Comment by aramperes** [comment]
Fixed in #854
